datafusion/common/src/logical_type/mod.rs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{fmt::Display, sync::Arc};
+
+use arrow_schema::{DataType, IntervalUnit, TimeUnit};
+
+use crate::logical_type::extension::{ExtensionType, ExtensionTypeRef};
+use crate::logical_type::field::{LogicalField, LogicalFieldRef};
+use crate::logical_type::fields::LogicalFields;
+
+pub mod type_signature;
+pub mod extension;
+pub mod registry;
+pub mod schema;
+pub mod field;
+pub mod fields;
+
+#[derive(Clone, Debug)]
+pub enum LogicalType {
```

Comment:
I generally agree with all the points discussed and this matches my intuition. The current state of the PR mostly reflects the state for the original proposal in #8143 as I didn't want to include any additional changes without discussing them first. I'm not familiar on how these changes should be discussed and if these points need to be also circulated in the mailing list.

Reply:
> There's probably a discussion to be had about what level we want logical type to be here. We previously discussed this, but I can't find that discussion anymore. (@alamb might remember.)

I also remember but can't find this discussion

As I recall, the conclusion was that types that can store different ranges of values shoudl be different logical types. 
* So that would mean `Int32` should be different than `UInt32`
* I don't recall consensus about String vs LargeString, but I agree that treating these as the same logical type with different phsical encodings would be the easiest to reason about (even though theoretically LargeString can represent strings larger than 2GB  that can not be represented in String)

Reply:
I've added support for treating large types as their _non-large_ logical type. I agree that types that can store different ranges of values should be different logical types